apply-license.man
.TH APPLY-LICENSE 1  "6 January 2011"
.\" Copyright © 2011 Bart Massey
.\" [This software is released under the "MIT License"]
.\" Please see the file COPYING in the source
.\" distribution of this software for license terms.
.SH NAME
apply-license \- apply a license in a directory
.SH SYNOPSIS
apply-license
.BI [ "<options> ..." ]
.BI [ "license_id" ]
.SH DESCRIPTION
.LP
The apply-license script is used to apply a license to the
source files in a directory.  The script will extract the
relevant information from a license file and apply it to any
source code it can find that does not already appear to have
a copyright notice.  Source code is identified by suffix.
Currently recognized suffixes are
".5c" (Nickle),
".awk",
".c",
".cabal",
".cc" (C++),
".cls" (LaTeX class),
".cpp" (C++),
".css",
".el" (Emacs Lisp),
".h",
".hs",
".html",
".java",
".js" (JavaScript),
".l" (Lex, not lisp),
".lisp" (Common Lisp),
".m" (Matlab/Octave),
".m4",
".man",
".php",
".pl",
".py",
".rb",
".ron" (Rust resource files),
".rs",
".scm" (Scheme),
".sh",
".sty" (LaTeX style),
".tex",
".y" (Yacc);
the files "Makefile"/"makefile" are also processed.
.LP
The first line of a script file may be a "#!" line and is
skipped if so.  The first line of a ".man" file may be a
".TH" line and is skipped if so.
.SH OPTIONS
.TP
.B "-q," "--quick"
This flag "quick-licenses" the project by creating a license
file from the given license id, adding a license note using
"--readme", and setting mode to "none".
.TP
.B "-m," "--mode " [ \fInone\fP | \fIlong\fP | \fIshort\fP | \fIheuristic\fP ]
Indicate the way in which the license will be applied to
source files. In
.I none
mode, the license will not be applied to files.  In
.I long
mode, the license will be applied in full at the end of
all files.  In
.I short
mode, a reference to the license file
will be applied.  In
.I heuristic
mode, each less than twice
the length of the license to be applied will be treated as
short. The default mode is
.IR short .
.TP
.B "-r"
This flag makes apply-license recurse into subdirectories.
.TP
.B "-u," "--username" "<\fIusername\fP>"
Full name of the user copyrighting the files. This is
required unless the environment variable
.B USERNAME
is present.
.TP
.B "-y," "--year" "<\fIyear\fP>"
Copyright year. Defaults to current year.
.TP
.B "-w," "--width" "<\fIwidth\fP>"
Text wrap width. Defaults to 64.
.TP
.B "--readme"
Edit the README to include the license. Chosen README is
README.tpl, README.md, README.txt, README in that order.
If no README exists, README.md will be created.
.TP
.B "--license-filename" "<\fIlicensefile\fP>"
Path to a file containing a license.  Default is LICENSE.txt
or LICENSE or COPYING (in that order) in the current
directory. If none exists, LICENSE.txt will be created.
The
.I "licensefile"
must consist of at least three paragraphs, with paragraphs
separated by a single blank line.  The first paragraph
should contain the copyright notice. The second paragraph
should contain a bracketed license name statement.
Subsequent paragraphs should contain the license text
itself.
.TP
.B "license_id"
Find a file named "license_id" in the license directory and
use that for the license. Incompatible with
.BR "--licensefile" .
.SH BUGS
.LP
Files without a suffix should be heuristically identified
and licensed if possible.
.LP
A file with a copyright already in it is silently
ignored. This can lead to problems if the cruddy
copyright-detection heuristic (currently "is the
word 'copyright' in there?") fails. See
http://github.com/BartMassey/emacs-copyright for an ugly
example.
.SH AUTHOR
Bart Massey <bart@cs.pdx.edu>


apply-license.py
#!/usr/bin/python
# Copyright © 2011 Bart Massey
# [This software is released under the "MIT License"]
# See the file COPYING in the source
# distribution of this software for license terms.

import argparse, datetime, os, re, sys, textwrap
from pathlib import Path

ap = argparse.ArgumentParser()
ap.add_argument("-q", "--quick", action="store_true")
ap.add_argument(
    "-m", "--mode",
    choices=["none", "long", "short", "heuristic"],
    default="short",
)
ap.add_argument("-l", "--license-filename", default="LICENSE.txt")
ap.add_argument("-r", "--recursive", action="store_true")
ap.add_argument("-u", "--username")
ap.add_argument("-y", "--year")
ap.add_argument("-w", "--width", type=int, default=64)
ap.add_argument("--readme", action="store_true")
ap.add_argument("--license-dir", default="/usr/local/lib/apply-license")
ap.add_argument("license_id", nargs="?")
args = ap.parse_args()

if args.quick:
    args.mode = "none"
    args.readme = True

def usage(mesg):
    print(f"apply-license: {mesg}")
    ap.print_usage()
    exit(1)

if args.username:
    username = args.username
elif "USERNAME" in os.environ:
    username = os.environ["USERNAME"]
else:
    usage("no username")

if args.year:
    year = args.year
else:
    year = datetime.date.today().year

def read_file(p, can_fail=False):
    try:
        with p.open("r") as f:
             return f.read().splitlines()
    except Exception as e:
        message = f"cannot read file: {e}"
        if can_fail:
            print(message)
            return None
        else:
            usage(message)

def write_file(p, text, can_fail=False):
    try:
        with p.open("w") as f:
             f.write('\n'.join(text) + '\n')
    except Exception as e:
        message = f"cannot write file: {e}"
        if can_fail:
            print(message)
            return
        else:
            usage(message)

def find_existing_license():
    if args.license_filename:
        targets = (args.license_filename,)
    else:
        targets = ("COPYING", "LICENSE", "LICENSE.txt")
    for license_file in targets:
        p = Path(license_file)
        if p.is_file():
            return p
    return None

def find_license():
    if not args.license_id:
        usage("no license specified")
        
    for suffix in ("-license.txt", ".txt", ""):
        lf = Path(args.license_dir) / (args.license_id + suffix)
        if lf.is_file():
            return lf

    usage(f"{args.license_id}: cannot find license file")

license_path = find_existing_license()
if license_path:
    if args.license_id:
        usage("license id specified but license file exists")
    known_license = False
    write_licensefile = False
    license_filename = str(license_path.name)
else:
    license_path = find_license()
    known_license = True
    license_filename = args.license_filename
    write_licensefile = True
license = read_file(license_path)

copyright = f"Copyright © {year} {username}"
if known_license:
    license_name = license[0].strip()
    release_statement = \
       f'This work is made available under the "{license_name}".'
else:
    stripped_license = [l for l in license if l.strip()]
    if re.search("[Cc]opyright", stripped_license[0]):
        copyright = stripped_license[0].strip()
        s = stripped_license[1].strip()
        info = re.fullmatch(r'\[([^"]*"([^"]+)"[^"]*)\]', s)
        if info:
            release_statement = info[1]
            if release_statement[-1] != ".":
                release_statement += "."
            license_name = info[2]
        else:
            license_name = "an open source license"
            release_statement = \
                "This work is made available under an open source license."

header_text = [
    copyright,
    "",
    f"[{release_statement}]",
    "",
]

if write_licensefile:
    new_license = header_text + license
    write_file(Path(license_filename), new_license)

def wrap(text):
    return textwrap.fill(' '.join(text), width = args.width).splitlines()

see_statement = [
    f'See the file {license_filename} in this distribution',
    'for license terms.',
]

see_statement_md = [
    f'See the file `{license_filename}` in this distribution',
    'for license terms.',
]

see_below_statement = [
    'See the end of this file for license terms.',
]

def update_readme(readme_path, md):
    if readme_path.exists():
        r = read_file(readme_path)
    else:
        r = []
    license_doc = [release_statement]
    if md:
        license_doc += see_statement_md
        r += ["", "## License"]
    else:
        license_doc += see_statement
    r += [""]
    r += wrap(license_doc)
    write_file(readme_path, r)

def find_and_update_readme():
    readmes = (
        ("README.tpl", True),
        ("README.md", True),
        ("README.txt", False),
        ("README", False),
    )
    for readme, md in readmes:
        readme_path = Path(readme)
        if readme_path.exists():
            update_readme(readme_path, md)
            return
    update_readme(Path("README.md"), True)

if args.readme:
    find_and_update_readme()

if args.mode == "none":
    exit(0)

def leading_comment(prefix):
    return lambda c: [prefix + (" " if l else "") + l for l in c]

def c_comment(c):
    star_comment = leading_comment(" *")
    return f"/*\n{star_comment(c)} */\n"

def html_comment(c):
    nop_comment = leading_comment("")
    return f"<!-- \n{nop_comment(c)}-->\n*/"

makefile_comment = leading_comment("#")

comment_styles = (
    (("c", "cc", "cpp", "h", "y", "l", "css", "java"), c_comment),
    (("html",), html_comment),
    (("hs", "cabal"), leading_comment("--")),
    (("js", "rust", "ron"), leading_comment("//")),
    (("lisp", "el", "scm"), leading_comment(";")),
    (("m4",), leading_comment("dnl")),
    (("tex", "cls", "sty"), leading_comment("%")),
    (
        ("sh", "bash", "awk", "5c", "rb", "pl", "py", "toml", "php"),
        leading_comment("#"),
    ),
    (("man",), leading_comment(r'.\"')),
)

commenters = {s : f for ss, f in comment_styles for s in ss}

def add_info(p):
    if p.name in ("Makefile", "makefile"):
        comment = commenters("sh")
    elif not p.suffix or p.suffix[1:] not in commenters:
        return
    else:
        comment = commenters[p.suffix[1:]]

    text = read_file(p, can_fail=True)
    if not text:
        return

    first_line = 0
    if p.suffix == ".man":
        if text[0].startswith(".TH"):
            first_line = 1
    elif text[0].startswith("#!"):
        first_line = 1

    head = '\n'.join(text[first_line:first_line + 3])
    if re.search("[Cc]opyright", head):
        return

    comment_text = list(header_text)
    mode = args.mode
    if mode == "heuristic":
        if len(text) < 2 * len(license):
            mode = "short"
        else:
            mode = "long"

    if mode == "short":
        comment_text += wrap(see_statement)
        text[first_line:first_line] = comment(comment_text)
    elif mode == "long":
        comment_text += wrap(see_below_statement)
        text[first_line:first_line] = comment(comment_text)
        text += ['']
        text += comment(license)

    write_file(p, text)

root = Path("")
if args.recursive:
    def exc(e):
        raise e
    paths = root.walk(on_error = exc)
else:
    paths = root.iterdir()

try:
    for p in paths:
        if p.is_file():
            add_info(p)
except Exception as e:
    usage(f"directory walk error: {e}")
